pulsar-client/src/main/java/org/apache/pulsar/client/impl/PartitionedProducerImpl.java

```diff
@@ -93,6 +98,19 @@ private MessageRouter getMessageRouter() {
 
         switch (messageRouteMode) {
             case CustomPartition:
+                String messageRouterClassName = conf.getMessageRouterClassName();
+                if (isNotBlank(messageRouterClassName)) {
+                    Class<?> clazz;
+                    try {
+                        clazz = Class.forName(messageRouterClassName);
+                        messageRouter = (MessageRouter) clazz.getDeclaredConstructor().newInstance();
+                        conf.setCustomMessageRouter(messageRouter);
+                    } catch (ClassNotFoundException | NoSuchMethodException | InvocationTargetException |
+                            InstantiationException | IllegalAccessException | ClassCastException e) {
+                        log.error("[{}] Error occurred while instantiating the messageRouterClassName",
```

Comment:
I believe that it's better to throw an exception rather than logging the exception. Please removing logging and throw the exception. It makes sense to wrap the original exception in PulsarClientException (with message "Error occurred while instantiating the messageRouterClassName " + messageRouterClassName and the original exception as the cause)

Reply:
It might not be easy to resolve the above suggestion since PulsarClientException is a checked exception and the the current methods don't current throw PulsarClientException. I'm not sure what wrapper exception to use. Perhaps you can check what would make sense.